src/rt/region/region_api.h

```diff
@@ -0,0 +1,332 @@
+    assert(
+      Region::get_type(r->get_region()) ==
+      Region::get_type(RegionContext::get_region()));
+
+    switch (Region::get_type(r->get_region()))
+    {
+      case RegionType::Trace:
+        RegionTrace::merge(ThreadAlloc::get(), RegionContext::get_entry_point(), r);
+        return r;
+      case RegionType::Arena:
+        RegionArena::merge(ThreadAlloc::get(), RegionContext::get_entry_point(), r);
+        return r;
+      case RegionType::Rc:
+        abort();
+    }
+  }
+
+  /**
+   * Create external reference to o in the current region.
+   */
+  inline ExternalRef* create_external_reference(Object* o)
+  {
+    return ExternalRef::create(RegionContext::get_region(), o);
+  }
+
+  /**
+   * Check if external reference is in the current region and still valid.
+   */
+  inline bool is_external_reference_valid(ExternalRef* e)
+  {
+    return e->is_in(RegionContext::get_region());
+  }
+
+  /**
+   * Use external reference e, and return the object in the current region.
+   */
+  inline Object* use_external_reference(ExternalRef* e)
+  {
+    assert(is_external_reference_valid(e));
+    return e->get();
+  }
+
+  /**
+   * Create object in current region
+   */
+  inline Object* create_object(Descriptor* d)
+  {
+    // Case analysis on type of region
+    switch (Region::get_type(RegionContext::get_region()))
+    {
+      case RegionType::Trace:
+        return RegionTrace::alloc(
+          ThreadAlloc::get(), RegionContext::get_entry_point(), d);
+      case RegionType::Arena:
+        return RegionArena::alloc(
+          ThreadAlloc::get(), RegionContext::get_entry_point(), d);
+      case RegionType::Rc:
+        return RegionRc::alloc(
+          ThreadAlloc::get(), (RegionRc*)RegionContext::get_region(), d);
+    }
+    // Unreachable as case is exhaustive
+    abort();
+  }
+
+  inline void add_reference(Object*)
+  {
+    // TODO
+  }
+
+  inline void remove_reference(Object*)
+  {
+    // TODO
+  }
+
+  inline void incref(Object* o)
+  {
+    RegionRc::incref(o);
+  }
+
+  inline void decref(Object* o)
+  {
+    switch (Region::get_type(RegionContext::get_region()))
+    {
+      case RegionType::Trace:
+        abort();
+        break;
+      case RegionType::Arena:
+        abort();
+        break;
+      case RegionType::Rc:
+        RegionRc::decref(ThreadAlloc::get(), o, (RegionRc*)RegionContext::get_region());
+        break;
+    }
+  }
+
+  template <typename T = Object>
+  inline T* create_fresh_region(RegionType type, Descriptor* d)
+  {
+    Object* entry_point = nullptr;
+    switch (type)
+    {
+      case RegionType::Trace:
+        entry_point = RegionTrace::create(ThreadAlloc::get(), d);
+        break;
+      case RegionType::Arena:
+        entry_point = RegionArena::create(ThreadAlloc::get(), d);
+        break;
+      case RegionType::Rc:
+        entry_point = RegionRc::create(ThreadAlloc::get(), d);
+        break;
+    }
+    return {reinterpret_cast<T*>(entry_point)};

```

Comment:
Why is this templatised to `T*`? What could `T` be other than `Object`?

Reply:
Can be a subtype of `Object*`, so then we can build a more type safe interface.